### DISCUSSION TITLE
How to obtain objective from optimize context?

### DISCUSSION THREAD

**Discussion by MohamedElgharbawy** [Z3Prover/z3#5647] [Q&A]
I am currently calling `Z3_optimize_from_string`, and I see in the [docs](https://z3prover.github.io/api/html/group__capi.html#gadde4c86439cb73e8357fb4edaa313c20) that it should:

> Add the parsed constraints and objectives to the optimization context.

However, I cannot seem to find what properties of the optimization context contain this information. Does anyone know where I can find this?

**Comment by NikolajBjorner** [comment]
For the C API, take a look at https://github.com/Z3Prover/z3/blob/af2cc460a96b6db97c774ef0b8926c145f26d216/src/api/z3_optimization.h#L340
There are two functions:

1. one function retrieves assertions
2. the other function retrieves objectives as a vector of expressions. Soft constraints are converted into a summation (+ (ite a 0 1) (ite b 0 1) ....)

**Reply by MohamedElgharbawy** [reply]
How can I get the values within the `AstVectorObj` that `Z3_optimize_get_objectives` returns? I tried indexing into it, but I got: `TypeError: 'AstVectorObj' object does not support indexing`.

**Reply by NikolajBjorner** [reply]
There are sufficient examples that illustrates accessing these objects in the examples directory.

**Reply by MohamedElgharbawy** [reply]
Sorry, I'm not seeing an example with an `AstVectorObj` in the examples directory. Could you point me in the right direction?

**Reply by MohamedElgharbawy** [reply]
I see the issue. I am working with Python, so I need a `z3.z3.AstVector` object, but the function that returns the objectives returns the C `AstVectorObj`. Any idea if I can convert it to the Python object?

**Reply by NikolajBjorner** [reply]
AstVectorObj is a python object. It contains iterators already and random access __get__). 
There are several examples in the python directory. you can also [t for t in opt.objectives()] to get a list.

**Reply by NikolajBjorner** [reply]
there is also help(AstVector) from python

**Reply by NikolajBjorner** [reply]
You don't call the C API directly from python. You should call the python functions.

**Reply by MohamedElgharbawy** [reply]
Then does there exist a python function for the C API's `Z3_optimize_get_objectives`? I could not find one.

**Reply by wintersteiger** [reply]
`Optimize::objectives()`?
